pyccel/ast/basic.py

```diff
@@ -23,13 +27,217 @@ def __new__(cls, *args, **kwargs):
+        replacement : object or tuple of objects
+                      The object which will be inserted instead
+        excluded_nodes : tuple of types
+                      Types for which substitute should not be called
+        """
+        if isinstance(original, tuple):
+            assert(isinstance(replacement, tuple))
+            assert(len(original) == len(replacement))
+        else:
+            original = (original,)
+            replacement = (replacement,)
+
+        for n in self._attribute_nodes:
+            v = getattr(self, n)
+            if v in original:
+                idx = original.index(v)
+                v.remove_user_node(self)
+                setattr(self, n, replacement[idx])
+                replacement[idx].user_nodes = self
+            elif isinstance(v, tuple):
+                new_v = []
+                for vi in v:
+                    if vi in original:
+                        idx = original.index(vi)
+                        vi.remove_user_node(self)
+                        new_v.append(replacement[idx])
+                        replacement[idx].user_nodes = self
+                    else:
+                        new_v.append(vi)
+                        vi.substitute(original, replacement, excluded_nodes)
+                setattr(self, n, v)
+            elif not isinstance(v, excluded_nodes):
+                v.substitute(original, replacement, excluded_nodes)
+
+    @property
+    def is_atomic(self):
+        """ Indicates whether the object has any attribute nodes
+        """
+        return bool(self._attribute_nodes)
+
     def set_fst(self, fst):
         """Sets the python.ast fst."""
-        self._fst = fst
+        if not isinstance(fst, ast.AST):
+            raise TypeError("Fst must be an AST object, not {}".format(type(fst)))
+        assert(isinstance(fst, ast.AST))
+
+        if not hasattr(fst, 'lineno'):
+            # Handle module object
+            fst.lineno     = 1
+            fst.col_offset = 1
+
+        self._fst.append(fst)
 
     @property
     def fst(self):
-        return self._fst
+        if len(self._fst) == 1:
+            return self._fst[0]
+        else:
+            return None
+
+    def get_direct_user_nodes(self, condition):
+        """ For an object with multiple user nodes
+        Get the objects which satisfy a given
+        condition
+
+        Parameters
+        ----------
+        condition : lambda
+                    The condition which the user nodes
+                    must satisfy to be returned
+        """
+        return [p for p in self._user_nodes if condition(p)]
+
+    @property
+    def user_nodes(self):
+        """ Get the current user_nodes of the object
+        """
+        if self._user_nodes:
+            return self._user_nodes[-1]
+        else:
+            return None
+
+    @user_nodes.setter
+    def user_nodes(self, user_nodes):
+        self._user_nodes.append(user_nodes)
```

Comment:
A find this setter strange: you will write `obj.user_nodes = node` instead of the more natural `obj.user_nodes += node`... So I recommend using a dedicated method called `set_current_user_node` (see comment above).

Reply:
I have renamed this `current_user_node` (9a1a43ee)

Reply:
`obj.user_nodes += node` is more natural when you remember that `user_nodes` is a list, but in almost all cases `user_nodes` only contains 1 object so I think it is more natural to interact with it this way (I think it should be all classes except, `Variable`, `FunctionDef` and `ClassDef`).
In practice when we want to interact with more than the current node we should not be interacting with the `_user_nodes` directly anyway

Reply:
All right